sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample11_MockingClientTypes.md

```diff
@@ -0,0 +1,456 @@
+            return eventCount < batchCountThreshold;
+        });
+
+// This sets up a mock of the CreateBatchAsync method.
+
+mockProducer
+    .Setup(p => p.CreateBatchAsync(It.IsAny<CancellationToken>()))
+    .ReturnsAsync(dataBatchMock);
+
+// Here we are mocking the SendAsync method so it will throw an exception if the batch passed into it is
+// not the one we are expecting to send.
+
+mockProducer
+    .Setup(p => p.SendAsync(
+        It.Is<EventDataBatch>(sendBatch => sendBatch != dataBatchMock),
+        It.IsAny<CancellationToken>()))
+    .Throws(new Exception("The batch published was not the expected batch."));
+
+EventHubProducerClient producer = mockProducer.Object;
+
+// The rest of this snippet demonstrates how to use the mocked methods define above. This would be where
+// application methods can be called and tested.
+
+EventDataBatch batch = await producer.CreateBatchAsync();
+
+// This is creating a list of events to use in our test.
+
+List<EventData> sourceEvents = new();
+
+for (int index = 0; index < batchCountThreshold; index++)
+{
+    var eventData = new EventData($"Sample-Event-{ index }");
+    sourceEvents.Add(eventData);
+}
+
+// Here we are adding events to the batch. They should all be accepted.
+
+foreach (var eventData in sourceEvents)
+{
+    Assert.True(batch.TryAdd(eventData));
+}
+
+// Since there area already batchCountThreshold number of events in the batch, this event will be rejected
+// from the batch.
+
+EventData eventData4 = new EventData("Sample-Event-4-will-fail");
+Assert.IsFalse(batch.TryAdd(eventData4));
+
+// For illustrative purposes we are calling SendAsync. This call would likely already be inside the
+// application-defined method being tested.
+
+await producer.SendAsync(batch);
+
+// This is demonstrating how to verify SendAsync was called once within the mocked producer client.
+
+mockProducer
+    .Verify(bp => bp.SendAsync(
+        It.IsAny<EventDataBatch>(),
+        It.IsAny<CancellationToken>()),
+    Times.Once);
+
+// For illustrative purposes, check that the events in the batch match what the application expects to have
+// added.
+
+foreach (EventData eventData in backingList)
+{
+    Assert.IsTrue(sourceEvents.Contains(eventData));
+}
+Assert.AreEqual(backingList.Count, sourceEvents.Count);
+```
+
+## Publishing events using an `EventHubBufferedProducerClient`
+
+The following snippet demonstrates how to mock the `EventHubBufferedProducerClient`. When using the buffered producer to publish events the  key interactions are calls to `EnqueueEventAsync` and/or `EnqueueEventsAsync`. Since the buffered producer implictly batches events passed in, mocking these calls is relatively straightforward.
+
+This snippet also demonstrates testing a send event failed handler. It is guaranteed by the library that the event handlers registered by the application will be called at the appropriate times. So fail and success handlers can be called and validated directly when testing.
```

Comment:
```suggestion
This snippet also demonstrates testing a send event failed handler. It is guaranteed by the library that the event handlers will be called at the appropriate times while the processor is running; it is recommended that handers be tested by calling them directly, rather than attempting to simulate having the buffered producer invoke them.
```